trl/trainer/ppo_trainer.py

```diff
@@ -1350,7 +1364,23 @@ def log_stats(
                     rewards = torch.tensor(rewards).to(self.current_device)
 
                 dist.barrier()
-                dist.all_reduce(rewards, op=torch.distributed.ReduceOp.SUM)
+                world_size = dist.get_world_size()
+                tensors = [torch.zeros_like(rewards) for _ in range(world_size)]
+                dist.all_gather(tensors, rewards)
+                rewards = torch.cat(tensors)
+
+                if self.config.log_with == "wandb":
+                    if any([column_to_log not in batch.keys() for column_to_log in columns_to_log]):
+                        raise ValueError(
+                            f"Columns to log {columns_to_log} are not present in the batch {batch.keys()}."
+                        )
+
+                    batch_list = [batch[column_to_log] for column_to_log in columns_to_log]
+
+                    dist.barrier()
+                    for b in batch_list:
+                        gather = [None for _ in range(dist.get_world_size())]
+                        dist.all_gather_object(gather, b)
```

Comment:
What is this doing?

Reply:
Here I am sending the query & response from process n>1 to process 0. This is two way, perhaps there is a way to do this one way with send / recv? I don't often use those operations.